PR: https://github.com/odoo/odoo/pull/169868

From: 238a41e35280256382f6509182b9e900fb4f7aba
From: Maxime de Neuville
Date: 2024-08-29 11:54:49

Breaking data model changes scores: del:9 + add:0, change matches:
-    _inherit = ['mrp.bom', 'analytic.mixin']
-    analytic_distribution_text = fields.Text(company_dependent=True)
-    analytic_distribution = fields.Json(inverse="_inverse_analytic_distribution", store=False, precompute=False)
-    analytic_account_ids = fields.Many2many('account.analytic.account', compute="_compute_analytic_account_ids", copy=True)
-    _inherit = ['mrp.production', 'analytic.mixin']
-    analytic_account_ids = fields.Many2many('account.analytic.account', compute='_compute_analytic_account_ids', store=True)
-    _inherit = 'stock.rule'

Total Changes: 130

[IMP] project,sale,purchase,stock,mrp: analytic plans on projects

PURPOSE

Currently, users can track the profitability of their projects, but they cannot allocate their timesheet-related costs to different entries. This approach is one-dimensional.
For instance, it can be relevant to track costs for the project as a whole, but also for the team or department in charge, the staff, a particular service, or a specific customer simultaneously.
By doing so, users can track the profitability of their company at a more granular level. They gain a better understanding of what is impacting their costs and can budget their projects accordingly.
Furthermore, we would like to handle the use-case of construction companies. Indeed, when stock is allocated to a construction site (or moved back to the warehouse if it wasn't used), it is not reflected in the profitability of the project because it doesn't generate analytic entries. This leads to an incorrect overview of the project's finances.
In addition, it is not easy to identify the stock moves impacting the project as there is no direct link between the two. It would be convenient for construction companies to have an overview and be able to create new pickings from the top bar of the project in their operations.
At last, these extra materials used sometimes need to be re-invoiced to the customer. Currently, this has to be done manually by adding a product to the SO. This is error-prone and time-consuming. We are thus relying on the 're-invoice expenses' feature that already exists for expenses and vendor bills to make it automatic.
Finally, another use-case that would benefit from these changes is companies that are engineering custom machines for their customers. They use projects to track the progress made on their prototypes but also need an overview of the POs, MOs, and stock pickings involved in the process.

FEATURE DETAILS

* Project
* Remove the 'analytic_account_id' field from projects
* Replace it with an analytic notebook allowing to define multiple analytic accounts (one per plan)
* Remove the analytic accounting for tasks
* Timesheets inherit the analytic accounts configured on their project or on the distribution of their associated SOL (priority to their SOL)

* Sale
* Remove the 'analytic_account_id' field from the SO
* Replace it with a project field. The analytic accounts of the project set on the SO are used to set a default SOL analytic distribution

* Stock
* Add an analytic costs field of operation types that indicates whether the stock picking of this type should generates AALs and do reinvoicing on validation
* Add a project field on stock pickings, used to generate AALs using the project's accounts and also for the reinvoicing of reinvoicable products
* Add a sales order field on projects to indicate on which SO to do the reinvoicing when validating stock pickings
* About the AAL generation, the amount is negative for operation of type 'delivery' and positive for type 'receipt'. If automatic accounting (anglo-saxon accounting) is enabled, do not generate AALs on stock picking validtion, to avoid generating in duplicate
* Landed costs's generated journal entries distribution receives the accounts of the stock picking's project or the MO's project that was selected

* MRP
* Remove the analytic distribution from MOs
* Replace it with a project field
* When generating AALs with the MO, again, the accounts are taken from the project linked to the MO
* Remove the analytic distribution from BOMs
* Replace it with a project field, used as a default value for setting the project of MOs

* Purchase
* Similarly to SOs, add a project field to POs
* Again, the accounts of the project linked to the PO are used as a default value for the distribution of the POLs

* Others
* Rename 'Project Updates' into 'Dashboard'
* Update the project profitability (report should reflect the analytic items linked to the analytic account of the 'Project' plan (in other words, we keep the current behavior))
* Update stat buttons and add top bars actions accordingly to the changes described above (SO, PO, From WH, To WH, MO, BOMs, Analytic Items)
* Adapt the reinvoicing of vendor bills to work with the project of the SO (and not the analytic account of the SO anymore)

task-3985258

closes odoo/odoo#169868

Related: odoo/enterprise#65054
Related: odoo/upgrade#6296
Signed-off-by: Xavier Bol (xbo) <xbo@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp_account/models/__init__.py
+++ b/addons/mrp_account/models/__init__.py
@@ -1,12 +1,9 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 from . import analytic_account
-from . import mrp_bom
 from . import mrp_workcenter
 from . import mrp_workorder
 from . import mrp_production
 from . import mrp_routing
 from . import product
 from . import stock_move
-from . import stock_rule
 from . import account_move

--- a/addons/mrp_account/models/mrp_bom.py
+++ b/None
@@ -1,42 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-import json
-
-from odoo import models, api, fields
-
-
-class MrpBom(models.Model):
-    _name = 'mrp.bom'
-    _inherit = ['mrp.bom', 'analytic.mixin']
-
-    # Company dependent JSON fields are not yet supported
-    analytic_distribution_text = fields.Text(company_dependent=True)
-    analytic_distribution = fields.Json(inverse="_inverse_analytic_distribution", store=False, precompute=False)
-    analytic_account_ids = fields.Many2many('account.analytic.account', compute="_compute_analytic_account_ids", copy=True)
-
-    @api.depends_context('company')
-    @api.depends('analytic_distribution_text')
-    def _compute_analytic_distribution(self):
-        for record in self:
-            record.analytic_distribution = json.loads(record.analytic_distribution_text or '{}')
-
-    def _inverse_analytic_distribution(self):
-        for record in self:
-            record.analytic_distribution_text = json.dumps(record.analytic_distribution)
-
-    @api.depends('analytic_distribution')
-    def _compute_analytic_account_ids(self):
-        for record in self:
-            record.analytic_account_ids = bool(record.analytic_distribution) and self.env['account.analytic.account'].browse(
-                list({int(account_id) for ids in record.analytic_distribution for account_id in ids.split(",")})
-            ).exists()
-
-    @api.onchange('product_id')
-    def _onchange_analytic_distribution(self):
-        for record in self:
-            if record.product_id:
-                record.analytic_distribution = record.env['account.analytic.distribution.model'].sudo()._get_distribution({
-                    "product_id": record.product_id.id,
-                    "product_categ_id": record.product_id.categ_id.id,
-                    "company_id": record.company_id.id,
-                })

--- a/addons/mrp_account/models/mrp_production.py
+++ b/addons/mrp_account/models/mrp_production.py
@@ -8,44 +8,15 @@ from odoo.tools import float_round
 
 
 class MrpProduction(models.Model):
-    _name = 'mrp.production'
-    _inherit = ['mrp.production', 'analytic.mixin']
+    _inherit = 'mrp.production'
 
     extra_cost = fields.Float(copy=False, string='Extra Unit Cost')
     show_valuation = fields.Boolean(compute='_compute_show_valuation')
-    analytic_account_ids = fields.Many2many('account.analytic.account', compute='_compute_analytic_account_ids', store=True)
 
     def _compute_show_valuation(self):
         for order in self:
             order.show_valuation = any(m.state == 'done' for m in order.move_finished_ids)
 
-    @api.depends('bom_id', 'product_id')
-    def _compute_analytic_distribution(self):
-        for record in self:
-            if record.bom_id.analytic_distribution:
-                record.analytic_distribution = record.bom_id.analytic_distribution
-            else:
-                record.analytic_distribution = record.env['account.analytic.distribution.model']._get_distribution({
-                    "product_id": record.product_id.id,
-                    "product_categ_id": record.product_id.categ_id.id,
-                    "company_id": record.company_id.id,
-                })
-
-    @api.depends('analytic_distribution')
-    def _compute_analytic_account_ids(self):
-        for record in self:
-            record.analytic_account_ids = bool(record.analytic_distribution) and self.env['account.analytic.account'].browse(
-                list({int(account_id) for ids in record.analytic_distribution for account_id in ids.split(",")})
-            ).exists()
-
-    @api.constrains('analytic_distribution')
-    def _check_analytic(self):
-        for record in self:
-            params = {'business_domain': 'manufacturing_order', 'company_id': record.company_id.id}
-            if record.product_id:
-                params['product'] = record.product_id.id
-            record.with_context({'validate_analytic': True})._validate_distribution(**params)
-
     def write(self, vals):
         res = super().write(vals)
         for production in self:
@@ -54,9 +25,6 @@ class MrpProduction(models.Model):
                 for workorder in production.workorder_ids:
                     workorder.mo_analytic_account_line_ids.ref = production.display_name
                     workorder.mo_analytic_account_line_ids.name = _("[WC] %s", workorder.display_name)
-            if 'analytic_distribution' in vals and production.state != 'draft':
-                production.move_raw_ids._account_analytic_entry_move()
-                production.workorder_ids._create_or_update_analytic_entry()
         return res
 
     def action_view_stock_valuation_layers(self):
@@ -69,16 +37,6 @@ class MrpProduction(models.Model):
         context['search_default_group_by_product_id'] = False
         return dict(action, domain=domain, context=context)
 
-    def action_view_analytic_accounts(self):
-        self.ensure_one()
-        return {
-            "type": "ir.actions.act_window",
-            "res_model": "account.analytic.account",
-            'domain': [('id', 'in', self.analytic_account_ids.ids)],
-            "name": _("Analytic Accounts"),
-            'view_mode': 'tree,form',
-        }
-
     def _cal_price(self, consumed_moves):
         """Set a price unit on the finished move according to `consumed_moves`.
         """

--- a/addons/mrp_account/models/mrp_workorder.py
+++ b/addons/mrp_account/models/mrp_workorder.py
@@ -39,17 +39,19 @@ class MrpWorkorder(models.Model):
         }
 
     def _create_or_update_analytic_entry(self):
-        for wo in self.filtered(lambda wo: wo.id and (wo.production_id.analytic_distribution or wo.workcenter_id.analytic_distribution or wo.wc_analytic_account_line_ids or wo.mo_analytic_account_line_ids)):
+        for wo in self:
+            if not wo.id:
+                continue
             hours = wo.duration / 60.0
             value = -hours * wo.workcenter_id.costs_hour
+            wo._create_or_update_analytic_entry_for_record(value, hours)
 
-            mo_analytic_line_vals = self.env['account.analytic.account']._perform_analytic_distribution(wo.production_id.analytic_distribution, value, hours, wo.mo_analytic_account_line_ids, wo)
-            if mo_analytic_line_vals:
-                wo.mo_analytic_account_line_ids += self.env['account.analytic.line'].sudo().create(mo_analytic_line_vals)
-
-            wc_analytic_line_vals = self.env['account.analytic.account']._perform_analytic_distribution(wo.workcenter_id.analytic_distribution, value, hours, wo.wc_analytic_account_line_ids, wo)
+    def _create_or_update_analytic_entry_for_record(self, value, hours):
+        self.ensure_one()
+        if self.workcenter_id.analytic_distribution or self.wc_analytic_account_line_ids or self.mo_analytic_account_line_ids:
+            wc_analytic_line_vals = self.env['account.analytic.account']._perform_analytic_distribution(self.workcenter_id.analytic_distribution, value, hours, self.wc_analytic_account_line_ids, self)
             if wc_analytic_line_vals:
-                wo.wc_analytic_account_line_ids += self.env['account.analytic.line'].sudo().create(wc_analytic_line_vals)
+                self.wc_analytic_account_line_ids += self.env['account.analytic.line'].sudo().create(wc_analytic_line_vals)
 
     def unlink(self):
         (self.mo_analytic_account_line_ids | self.wc_analytic_account_line_ids).unlink()

--- a/addons/mrp_account/models/stock_move.py
+++ b/addons/mrp_account/models/stock_move.py
@@ -14,18 +14,12 @@ class StockMove(models.Model):
         res += self.filtered(lambda m: m.bom_line_id.bom_id.product_tmpl_id.id == product.product_tmpl_id.id)
         return res
 
-    def _get_analytic_distribution(self):
-        distribution = self.raw_material_production_id.analytic_distribution
-        if distribution:
-            return distribution
-        return super()._get_analytic_distribution()
-
     def _should_force_price_unit(self):
         self.ensure_one()
         return self.picking_type_id.code == 'mrp_operation' or super()._should_force_price_unit()
 
     def _ignore_automatic_valuation(self):
-        return bool(self.raw_material_production_id)
+        return super()._ignore_automatic_valuation() or bool(self.raw_material_production_id)
 
     def _get_src_account(self, accounts_data):
         if self._is_production():

--- a/addons/mrp_account/models/stock_rule.py
+++ b/None
@@ -1,17 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models
-
-
-class StockRule(models.Model):
-    _inherit = 'stock.rule'
-
-    def _prepare_mo_vals(self, product_id, product_qty, product_uom, location_id, name, origin, company_id, values, bom):
-        res = super()._prepare_mo_vals(product_id, product_qty, product_uom, location_id, name, origin, company_id, values, bom)
-        if not bom.analytic_distribution:
-            if values.get('analytic_distribution'):
-                res['analytic_distribution'] = values.get('analytic_distribution')
-            elif values.get('analytic_account_id'):
-                res['analytic_distribution'] = {values.get('analytic_account_id').id: 100}
-        return res
